### PR TITLE
feat: port `GQSP` to `Operator2`

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**
Migrating `GQSP` to use `Operator2` as its base.

**Description of the Change:**
* Migrate `GQSP`
* Add `skip_bind_new_parameters` to `assert_valid` to fix some test failures, mimicking #9814. `bind_new_parameters`'s fate is being discussed right now wrt to `Operator2`, so this change should hopefully not be very controversial.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-125644]

### CODE REVIEW

File: pennylane/ops/functions/assert_valid.py [hunk 1]
```diff
@@ -174,9 +174,10 @@ def _check_decomposition_new(op, skip_decomp_matrix_check=False):
     for rule in qp.list_decomps(f"C({op.name})"):
         for n_ctrl_wires, c_value, n_workers in itertools.product([1, 2, 3], [0, 1], [0, 1, 2]):
             ctrl = qp.ops.Controlled if isinstance(op, Operator1) else qp.ops.ControlledOp2
-            control_wires = [i + len(op.wires) for i in range(n_ctrl_wires)]
+            max_wire = max(op.wires) if op.wires else 0
+            control_wires = [i + max_wire + 1 for i in range(n_ctrl_wires)]
             control_values = [c_value] * n_ctrl_wires
-            work_wires = [i + len(op.wires) + n_ctrl_wires for i in range(n_workers)]
+            work_wires = [i + max(control_wires) + 1 for i in range(n_workers)]
```
Comment (mudit2812) [inline pennylane/ops/functions/assert_valid.py:184]: This change was necessary because for hybrid operators, there could be situations where the wires weren't equal to `range(num_wires)` as the inner operator wouldn't follow that convention, only the outer operator. The only downside of this is that we cannot use assert_valid with operators that have non-int wire labels, but we're getting rid of that anyway

File: pennylane/ops/functions/assert_valid.py [hunk 2]
```diff
@@ -646,6 +648,7 @@ def _assert_valid_operator2(
         skip_pickle: If ``True``, the pickle test will be skipped.
         skip_wire_mapping: If ``True``, the wire mapping test will be skipped.
         skip_capture: If ``True``, the program capture test will be skipped.
+        skip_bind_new_parameters: If ``True``, the ``bind_new_parameters`` test will be skipped.
```
Comment (mudit2812) [inline pennylane/ops/functions/assert_valid.py:680]: Mimicking the change made by Astral in #9814

File: pennylane/estimator/resource_mapping.py [hunk 3]
```diff
@@ -534,10 +534,10 @@ def _(op: qtemps.Reflection):
 
 @_map_to_resource_op.register
 def _(op: qtemps.GQSP):
-    be_op = op.hyperparameters["unitary"]
+    be_op = op.arguments["unitary"]
```
Comment (astralcai) [inline pennylane/estimator/resource_mapping.py]:
```suggestion
    be_op = op.unitary
```
Would this work as well?

Comment (mudit2812) [reply pennylane/estimator/resource_mapping.py]: Yes. I used `arguments` before due to all the `no-member` errors, but that should no longer be an issue.

File: pennylane/estimator/resource_mapping.py [hunk 4]
```diff
@@ -534,10 +534,10 @@ def _(op: qtemps.Reflection):
 
 @_map_to_resource_op.register
 def _(op: qtemps.GQSP):
-    be_op = op.hyperparameters["unitary"]
+    be_op = op.unitary
     mapped_be_op = _map_to_resource_op(be_op)
 
-    ctrl_wire = op.hyperparameters["control"]
+    ctrl_wire = op.arguments["control"]
```
Comment (astralcai) [inline pennylane/estimator/resource_mapping.py]:
```suggestion
    ctrl_wire = op.control
```

File: pennylane/templates/subroutines/gqsp.py [hunk 5]
```diff
@@ -88,101 +89,38 @@ def circuit(angles):
...
         new_op._wires = Wires([wire_map.get(wire, wire) for wire in self.wires])
-        new_op._hyperparameters["unitary"] = ops.functions.map_wires(
-            new_op._hyperparameters["unitary"], wire_map
-        )
-        new_op._hyperparameters["control"] = tuple(
-            wire_map.get(w, w) for w in Wires(new_op._hyperparameters["control"])
+        new_op.arguments["unitary"] = ops.functions.map_wires(new_op.arguments["unitary"], wire_map)
+        new_op.arguments["control"] = tuple(
+            wire_map.get(w, w) for w in Wires(new_op.arguments["control"])
         )
```
Comment (astralcai) [inline pennylane/templates/subroutines/gqsp.py]: Does the automatic `Operator2.map_wires` not work here?

File: pennylane/templates/subroutines/gqsp.py [hunk 6]
```diff
@@ -88,101 +89,38 @@ def circuit(angles):
...
-            op_list.append(ops.X(control))
-            op_list.append(ops.Z(control))
-
-        return op_list
-
+    @override
     def queue(self, context=QueuingManager):
-        context.remove(self.hyperparameters["unitary"])
+        context.remove(self.arguments["unitary"])
         context.append(self)
```
Comment (astralcai) [inline pennylane/templates/subroutines/gqsp.py]: Does `Operator2.queue` automatically remove hybrid args from queue?

File: pennylane/templates/subroutines/gqsp.py [hunk 7]
```diff
@@ -88,101 +89,38 @@ def circuit(angles):
...
-        # Data contains (angles, derived_data_from_unitary, unitary)
-        return cls(angles=data[0], unitary=data[-1], control=metadata[0])
-
-    # pylint: disable=arguments-differ
-    @classmethod
-    def _primitive_bind_call(cls, unitary, angles, control):
-        return super()._primitive_bind_call(unitary, angles, wires=control)
+    @override
+    def wires(self):
+        return self.arguments["control"] + self.arguments["unitary"].wires
```
Comment (astralcai) [inline pennylane/templates/subroutines/gqsp.py]:
```suggestion
        return self.control + self.unitary.wires
```

File: pennylane/templates/subroutines/gqsp.py (hunk 6 above)
Comment (mudit2812) [reply pennylane/templates/subroutines/gqsp.py]: It doesn't, but it probably should

File: pennylane/templates/subroutines/gqsp.py (hunk 5 above)
Comment (mudit2812) [reply pennylane/templates/subroutines/gqsp.py]: It likely will and this can probably be removed. Let me try

Comment (mudit2812) [reply pennylane/templates/subroutines/gqsp.py]: Works

File: pennylane/templates/subroutines/gqsp.py (hunk 6 above)
Comment (mudit2812) [reply pennylane/templates/subroutines/gqsp.py]: After making a small update to `Operator2.queue`, we no longer need to override `GQSP.queue`

**Review by astralcai — APPROVED** [review]
Nice!

File: pennylane/templates/subroutines/gqsp.py [hunk 8]
```diff
@@ -88,101 +89,21 @@ def circuit(angles):
...
-        Args:
-            *parameters (list): trainable parameters of the operator, as stored in the ``parameters`` attribute
-            wires (Iterable[Any], Wires): wires that the operator acts on
-            **hyperparams (dict): non-trainable hyperparameters of the operator, as stored in the ``hyperparameters`` attribute
-
-        Returns:
-            list[Operator]: decomposition of the operator
-        """
-
-        unitary = hyperparameters["unitary"]
```
Comment (comp-phys-marc) [inline pennylane/templates/subroutines/gqsp.py:153]: Really nice that we can get rid of all of this :)

File: tests/capture/test_templates.py [hunk 9]
```diff
@@ -1595,37 +1595,33 @@ def qfunc():
...
 
         angles = np.ones([3, 3])
-        unitary = qp.RX(1, wires=1)
+        unitary = qp.S(wires=1)
         # Validate inputs
         qfunc(unitary, angles)
 
         # Actually test primitive bind
         jaxpr = jax.make_jaxpr(qfunc)(unitary, angles)
+        print(jaxpr)
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: We probably don't want this print statement in here

File: tests/capture/test_templates.py [hunk 10]
```diff
@@ -1595,37 +1595,33 @@ def qfunc():
...
+        # Dynamic args first, then wires, then hybrid args, so first arg will be the angles,
+        # then the control wire, then the wire of the S gate
+        assert gqsp_eqn.invars[0] == jaxpr.jaxpr.invars[1]  # Angles
+        assert gqsp_eqn.invars[1].val == 0  # Control wire
+        assert gqsp_eqn.invars[2] == jaxpr.jaxpr.invars[0]  # S wire
 
-        rx_eqn = jaxpr.eqns[0]
-        assert rx_eqn.primitive == qp.RX._primitive
-        gqps_eqn = jaxpr.eqns[1]
-        assert gqps_eqn.primitive == qp.GQSP._primitive
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py:1616]: We could replace this with a use of `assert_eqn_matches_op`

**Review by comp-phys-marc — COMMENTED** [review]
Just two minor requests, otherwise looking great!

File: tests/capture/test_templates.py (hunk 10 above)
Comment (mudit2812) [reply tests/capture/test_templates.py:1616]: I already did this, but it was quite recent so you might've reviewed an older commit 😅

File: tests/capture/test_templates.py (hunk 9 above)
Comment (mudit2812) [reply tests/capture/test_templates.py]:
```suggestion
```

File: tests/capture/test_templates.py (hunk 10 above)
Comment (comp-phys-marc) [reply tests/capture/test_templates.py:1616]: Oh yeah I see it now thanks!

**Review by comp-phys-marc — APPROVED** [review]
Very nice!
